### ISSUE TITLE
Thread safety in HttpProtocol when concurrent = true

### ISSUE THREAD

**Issue by fmontesi** [jolie/jolie#491]
In `HttpProtocol`, we have the following thread-unsafe fields:

```
	// FIXME the following fields need to be made thread-safe
	private String inputId = null;
	private MultiPartFormDataParser multiPartFormDataParser = null;
	private String encoding = null;
	private String requestFormat = null;
	private String requestCharset = null;
	private boolean headRequest = false;
```

This is a problem when `isThreadSafe() == true` because of data races when dealing with concurrent HTTP requests.

@mwallnoefer would it be possible to at least group all of these parameters into an enclosing object (a private `HttpExchangeContext` class or something like that?). That way we can start either:
- Deactivating content negotiation & co. when `isThreadSafe() == true`; or, even better,
- Have a concurrent hashmap to keep track of these context objects per request id.

**Comment by mwallnoefer** [comment]
Branch https://github.com/jolie/jolie/tree/fix/http-threadsafety

**Comment by mwallnoefer** [comment]
In the _sodep_ protocol we have the same problem with the _charset_ parameter (field `stringCharset`).

**Comment by mwallnoefer** [comment]
@fmontesi In the feature branch I have started with the refactoring for the concurrent hash map, but the server keeps crashing.

**Comment by fmontesi** [comment]
Yeah, this is complicated to handle, and we also risk memory leaks in case we forget to remove things from these maps (and that's hard to do correctly).

I was wondering if we could instead allow to extend CommMessage to contain metadata that gets copied to response messages generated by the interpreter. This way we could retrieve metadata about the original request in http protocol in a safe way (accessing it from the message we're sending back). What do you think?

Otherwise we could switch to a concurrent hash map that maps request IDs to entire context objects and set some timeouts I guess.

**Comment by fmontesi** [comment]
Alright, I've played a bit with the idea. I've created a framework for attaching metadata to messages here:

https://github.com/jolie/jolie/tree/feature/message-metadata

We can use it as follows (next steps):
- Update `CommMessage.createResponse` such that the created CommMessage stores a reference to the original request. This would enable you to obtain the metadata of the original request when you send a response in the http extension.
- Store the `ExchangeContext` as metadata inside of the CommMessage we create when receiving a message in http.
- Access the `ExchangeContext` (actually now we could make it a far more general data structure, but one thing at a time... :-)) of the original request from within the send procedure of the http protocol.

What do you think @mwallnoefer ?

**Comment by mwallnoefer** [comment]
I'm deeply impressed 🤩 . Unfortunately I lack the insight about the interactions between Jolie's internals (like the messaging subsystem), so it is a bit hard to follow for me. But the approach on behalf of metadata annotations sounds very good 😃 .

**Comment by fmontesi** [comment]
Looking good so far: https://github.com/jolie/jolie/tree/feature/message-metadata

I've switched `requestCharset` and `requestFormat` to the new system. Once we've moved all of them, we can get rid of the `exchangeContext` field in HTTPProtocol (we'll still need the type, that's what I've used to store the http-related metadata in messages).

@mwallnoefer could you try porting the other fields as well? (inputId, multiPartFormDataParser, etc.) Would be a good way of seeing if the new system is intuitive enough. :-)

**Comment by mwallnoefer** [comment]
Thanks, @fmontesi.

So I will fetch the branch and add `inputId`, `multiPartFormDataParser`, `encoding` and `headRequest`?

**Comment by fmontesi** [comment]
Yes, thanks!

About the commented code: yeah, I think we can just remove it. I initially made a fancy mechanism that checked that metadata keys were unique. But Jolie extensions can actually be loaded multiple times by different classloaders, thus static fields are not actually unique. So I scrapped it.

**Comment by mwallnoefer** [comment]
I've added my commit, please review it (hopefully I've done it right)! And I do not need to reset anything to `= false` or ` = null` since the old attribute contexts get invalidated automatically, correct?  

But as a last showstopper I do not know how to get the request's `inputId` in the receive part, since I have no appropriate `CommMessage` object there.

**Comment by fmontesi** [comment]
Right, that's trickier! Now done. :⁠-⁠)

There are a few tests that fail because of status codes, which seem unrelated to this issue. Do you agree? Have they been caused by the recent changes to status codes maybe?

I haven't noticed because, apparently, even if these tests fail the GitHub action thinks everything went alright. That's another issue I guess..

**Comment by mwallnoefer** [comment]
Well done!

But it is strange that now some tests start to fail. I mean `mvn install` respectively `jolie test.ol` should pass without errors 🤔 .

**Comment by mwallnoefer** [comment]
The custom exception patch is the last showstopper: https://github.com/jolie/jolie/commit/d77bf8225ebc656c1282516bd0f778ce1a216a27. Any idea how to fix that one?

**Comment by fmontesi** [comment]
All fixed now I think. :-) Closing this for now.

**Comment by mwallnoefer** [comment]
Fabrizio, do you plan to enable it per default e.g. on Leonardo? Maybe the public Jolie website could become the testing environment of this feature? Just to see how stable it is...

**Comment by fmontesi** [comment]
Enable what by default? Message metadata is now gonna be used transparently
under the hood. Or are you talking about `concurrent = true`? That works
only for Jolie to Jolie interactions anyway. Unless other techs start
picking up our custom HTTP headers... :-)

**Comment by mwallnoefer** [comment]
I guess we have some remaining issue here (HttpProtocol.java, `operationName` is not set.
```
       private void recv_parseMessage( HttpMessage message, HttpUtils.DecodedMessage deco>
                String charset, ExchangeContext messageMetadata )
                throws IOException {
                final String operationName;
                if( decodedMessage.operationName != null ) {
                        operationName = decodedMessage.operationName;
                } else if( message.isResponse() ) {
                        operationName = inputId;
                } else {
                        operationName = null;
                }

                if( getOperationSpecificStringParameter( operationName, HttpUtils.Paramete>
                        .equals( NativeType.STRING.id() ) ) {
                        decodedMessage.value.setValue( new String( message.content(), char>
```

Stacktrace:
```
get-stations:Exception in thread "Thread-1" java.lang.NullPointerException
	at java.base/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936)
	at java.base/java.util.concurrent.ConcurrentHashMap.containsKey(ConcurrentHashMap.java:964)
	at jolie.runtime.ValueImpl.hasChildren(Value.java:183)
	at jolie.net.protocols.CommProtocol.getOperationSpecificStringParameter(CommProtocol.java:182)
	at jolie.net.HttpProtocol.recv_parseMessage(HttpProtocol.java:1132)
	at jolie.net.HttpProtocol.recv_internal(HttpProtocol.java:1405)
	at jolie.net.http.HttpUtils.recv(HttpUtils.java:324)
	at jolie.net.HttpProtocol.recv(HttpProtocol.java:1
[odh2.ol.txt](https://github.com/jolie/jolie/files/15293277/odh2.ol.txt)
527)
...
```

I have attached you some small code sample.

**Comment by mwallnoefer** [comment]
Fixed by this commit, please merge: https://github.com/jolie/jolie/pull/505/commits/79b1f278a6a0d4e3c30095b66ac6d14260cc332e
